src/ucp/tag/eager_rcv.c

```diff
@@ -150,16 +150,26 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_eager_first_handler,
 }
 
 static UCS_F_ALWAYS_INLINE ucs_status_t
-ucp_eager_common_middle_handler(ucp_worker_t *worker, ucp_tag_frag_match_t *matchq,
-                                khiter_t iter, void *data, size_t length,
+ucp_eager_common_middle_handler(ucp_worker_t *worker, void *data, size_t length,
                                 uint16_t hdr_len, unsigned tl_flags,
                                 uint16_t flags, uint16_t priv_length)
 {
     ucp_eager_middle_hdr_t *hdr = data;
+    ucp_tag_frag_match_t *matchq;
     ucp_recv_desc_t *rdesc;
     ucp_request_t *req;
     ucs_status_t status;
     size_t recv_len;
+    khiter_t iter;
+    int ret;
+
+    iter   = kh_put(ucp_tag_frag_hash, &worker->tm.frag_hash, hdr->msg_id, &ret);
+    ucs_assert(ret >= 0);
+    matchq = &kh_value(&worker->tm.frag_hash, iter);
+    if (ret != 0) {
+        /* initialize a previously empty hash entry */
+        ucp_tag_frag_match_init_unexp(matchq);
+    }
```

Comment:
- ucs_assert(ret == 0) in 182
 - remove kh elem in 'else' of the 'if' in line 178, and queue_is_empty
 - add some assertion in the end of the function that the element is not empty unexp. add comment that we cannot allow empty unexp elem because it points to itself which is not valid for khash elem